master/pkg/tasks/trial.go

```diff
@@ -0,0 +1,231 @@
+		s.Rank,
+	)
+}
+
+// Entrypoint implements InnerSpec.
+func (s StartTrial) Entrypoint() []string {
+	return []string{"/run/determined/train/entrypoint.sh"}
+}
+
+// Environment implements InnerSpec.
+func (s StartTrial) Environment(t TaskSpec) expconf.EnvironmentConfig {
+	env := s.ExperimentConfig.Environment()
+	ports := env.Ports()
+	if ports == nil {
+		ports = make(map[string]int)
+	}
+	ports["trial"] = rendezvousPort(trialUniquePortOffset(t.Devices))
+	env.SetPorts(ports)
+	return env
+}
+
+// EnvVars implements InnerSpec.
+func (s StartTrial) EnvVars(t TaskSpec) map[string]string {
+	portOffset := trialUniquePortOffset(t.Devices)
+	portStr := rendezvousPort(portOffset)
+	return map[string]string{
+		"DET_EXPERIMENT_ID":            fmt.Sprintf("%d", s.InitialWorkload.ExperimentID),
+		"DET_TRIAL_ID":                 fmt.Sprintf("%d", s.InitialWorkload.TrialID),
+		"DET_TRIAL_SEED":               fmt.Sprintf("%d", s.TrialSeed),
+		"DET_EXPERIMENT_CONFIG":        jsonify(s.ExperimentConfig),
+		"DET_HPARAMS":                  jsonify(s.HParams),
+		"DET_INITIAL_WORKLOAD":         jsonify(s.InitialWorkload),
+		"DET_LATEST_CHECKPOINT":        "/run/determined/train/checkpoint.json",
+		"DET_WORKLOAD_MANAGER_TYPE":    string(s.WorkloadManagerType),
+		"DET_RENDEZVOUS_PORT":          strconv.Itoa(portStr),
+		"DET_TRIAL_UNIQUE_PORT_OFFSET": strconv.Itoa(portOffset),
+	}
+}
+
+// LoggingFields implements InnerSpec.
+func (s StartTrial) LoggingFields() map[string]string {
+	return map[string]string{
+		"trial_id": strconv.Itoa(s.InitialWorkload.TrialID),
+	}
+}
+
+// Mounts implements InnerSpec.
+func (s StartTrial) Mounts() []mount.Mount {
+	mounts := ToDockerMounts(s.ExperimentConfig.BindMounts())
+	addMount := func(source, target string, bindOpts *mount.BindOptions) {
+		mounts = append(mounts, mount.Mount{
+			Type: mount.TypeBind, Source: source, Target: target, BindOptions: bindOpts,
+		})
+	}
+
+	if c := s.ExperimentConfig.CheckpointStorage().RawSharedFSConfig; c != nil {
+		addMount(
+			c.HostPath(),
+			model.DefaultSharedFSContainerPath,
+			&mount.BindOptions{Propagation: model.DefaultSharedFSPropagation},
+		)
+	}
+
+	if c := s.ExperimentConfig.DataLayer().RawSharedFSConfig; c != nil {
+		if c.HostStoragePath() != nil && c.ContainerStoragePath() != nil {
+			addMount(*c.HostStoragePath(), *c.ContainerStoragePath(), nil)
+		}
+	}
+	if c := s.ExperimentConfig.DataLayer().RawS3Config; c != nil {
+		if c.LocalCacheHostPath() != nil && c.LocalCacheContainerPath() != nil {
+			addMount(*c.LocalCacheHostPath(), *c.LocalCacheContainerPath(), nil)
+		}
+	}
+	if c := s.ExperimentConfig.DataLayer().RawGCSConfig; c != nil {
+		if c.LocalCacheHostPath() != nil && c.LocalCacheContainerPath() != nil {
+			addMount(*c.LocalCacheHostPath(), *c.LocalCacheContainerPath(), nil)
+		}
+	}
+
+	return mounts
+}
+
+// UseFluentLogging implements InnerSpec.
+func (s StartTrial) UseFluentLogging() bool { return true }
+
+// UseHostMode implements InnerSpec.
```

Comment:
Could we rename `UseHostMode` to be `ForceHostMode`?  That is what it actually means.

Reply:
Why do you need to force it? It seems like just using it would be effective

Reply:
Because the value of `UseHostMode` is only used when it is true.  That is why `ForceHostMode` is a better name for what it actually does.

Reply:
This should apply to other fields. like `UseFluentLogging`. It can also be `ForceFluentLogging`.

Reply:
This is what I'm trying to say:
```go
        // master/pkg/task/task.go::ToContainerSpec()
	network := t.TaskContainerDefaults.NetworkMode
	if t.UseHostMode() {
		network = hostMode
	}
```

That is, `UseHostMode` doesn't mean that we "use host mode".  It means that we _override the host mode setting from the task container defaults_.  If it was called `ForceHostMode()` it would be a better name.

That's not true for `UseFluentLogging()`.  UseFluentLogging literally returns if we should use fluent logging or not:
```
	spec := container.Spec{
		// ...
		RunSpec: container.RunSpec{
			// ...
			UseFluentLogging: t.UseFluentLogging(),
		},
	}